src/peers.py

```diff
@@ -130,105 +130,193 @@ def peerline(cls, line):
+        '.LOCL.',
+        '.POOL.',
+        '.STEP.',
+        '.XFAC.',
+    )
+
+    @classmethod
+    def ntpd_peerline(cls, tally, fields):
+        """Convert ntpd peer line to named fields"""
+        if len(fields) != 10:
+            return None
+        if fields[1] in cls.ignorerefids:
+            return None
+        return {
+            'tally': tally,
+            'address': fields[0],
+            'refid': fields[1],
+            'stratum': fields[2],
+            'mode': fields[3],
+            'when': fields[4],
+            'poll': fields[5],
+            'reach': fields[6],
+            'delay': fields[7],
+            'offset': fields[8],
+            'jitter': fields[9],
+        }
+
+    @classmethod
+    def validate_tally(cls, fields):
+        peertype = cls.tallytotype(fields['tally'])
+        if peertype == 'unknown':
+            # Unknown peer tally code
             return False
+        else:
+            fields['tally'] = peertype
+        return True
 
-        # stratum should be an integer
+    @classmethod
+    def validate_stratum(cls, fields):
+        # stratum should be an integer between 0 & 15
         try:
-            fields[2] = int(fields[2])
+            fields['stratum'] = int(fields['stratum'])
+            if fields['stratum'] < 0 or fields['stratum'] > 15:
+                # stratum out of bounds
+                return False
         except ValueError:
-            warn('Stratum is not an integer: %s' % fields[2])
+            # stratum not an integer
             return False
+        return True
 
+    @classmethod
+    def validate_when(cls, fields):
         # when should be an integer or '-'
         try:
-            if fields[4] != '-':
-                fields[4] = NTPPeers.time2seconds(fields[4])
+            if fields['when'] != '-':
+                fields['when'] = NTPPeers.time2seconds(fields['when'])
         except ValueError:
-            warn('Last poll time is not an integer: %s' % fields[4])
+            # last poll time not an integer
             return False
+        return True
 
-        # poll should be an integer
+    @classmethod
+    def validate_poll(cls, fields):
+        # poll should be an integer in seconds
+        if 'poll_pow2' in fields:
+            try:
+                fields['poll'] = 2 ** int(fields['poll_pow2'])
+            except ValueError:
+                # poll interval not an integer
+                return False
         try:
-            fields[5] = int(fields[5])
+            fields['poll'] = int(fields['poll'])
         except ValueError:
-            warn('Poll interval is not an integer: %s' % fields[5])
+            # poll interval not an integer
             return False
+        return True
 
+    @classmethod
+    def validate_reach(cls, fields):
         # reachability should be octal
         try:
             # convert to octal
-            fields[6] = int(fields[6], 8)
+            fields['reach'] = int(fields['reach'], 8)
```

Comment:
This is converting _from_ octal, the opposite of what the comment says.

Reply:
Fixed comment